src/yunohost/user.py

```diff
@@ -566,6 +607,225 @@ def user_info(username):
+                                line=reader.line_num,
+                                details=', '.join(format_errors)))
+            is_well_formatted = False
+            continue
+
+        # Choose what to do with this line and prepare data
+        user['groups'] = to_list(user['groups'])
+        user['mail-alias'] = to_list(user['mail-alias'])
+        user['mail-forward'] = to_list(user['mail-forward'])
+        user['domain'] = user['mail'].split('@')[1]
+        if user['username'] not in existing_users:
+            # Generate password if not exists
+            # This could be used when reset password will be merged
+            if not user['password']:
+                user['password'] = random_ascii(70)
+            actions['created'].append(user)
+        else:
+            if update:
+                actions['updated'].append(user)
+            del existing_users[user['username']]
+
+    if delete:
+        for user in existing_users:
+            actions['deleted'].append(user)
+
+    if not is_well_formatted:
+        raise YunohostError('user_import_bad_file')
+
+    total = len(actions['created'] + actions['updated'] + actions['deleted'])
+
+    if total == 0:
+        logger.info(m18n.n('user_import_nothing_to_do'))
+        return
+
+    # Apply creation, update and deletion operation
+    result = {
+        'created': 0,
+        'updated': 0,
+        'deleted': 0,
+        'errors': 0
+    }
+
+    def progress(info=""):
+        progress.nb += 1
+        width = 20
+        bar = int(progress.nb * width / total)
+        bar = "[" + "#" * bar + "." * (width - bar) + "]"
+        if info:
+            bar += " > " + info
+        if progress.old == bar:
+            return
+        progress.old = bar
+        logger.info(bar)
+    progress.nb = 0
+    progress.old = ""
+
+    def on_failure(user, exception):
+        result['errors'] += 1
+        logger.error(user + ': ' + str(exception))
+
+    def update(user, info=False):
+        remove_alias = None
+        remove_forward = None
+        if info:
+            user['mail'] = None if info['mail'] == user['mail'] else user['mail']
+            remove_alias = list(set(info['mail-alias']) - set(user['mail-alias']))
+            remove_forward = list(set(info['mail-forward']) - set(user['mail-forward']))
+            user['mail-alias'] = list(set(user['mail-alias']) - set(info['mail-alias']))
+            user['mail-forward'] = list(set(user['mail-forward']) - set(info['mail-forward']))
+            for group, infos in user_group_list()["groups"].items():
+                if group == "all_users":
+                    continue
+                # If the user is in this group (and it's not the primary group),
+                # remove the member from the group
+                if user['username'] != group and user['username'] in infos["members"]:
+                    user_group_update(group, remove=user['username'], sync_perm=False, imported=True)
```

Comment:
To be re-discussed, maybe we can optimize this to not spend a huge time adding/removing from/to groups